tensor/indexed.go

```diff
@@ -0,0 +1,706 @@
+// which is the length of the index list or number of outer
+// rows dimension of tensor if no indexes (full sequential view).
+func (ix *Indexed) NumRows() int {
+	if ix.Indexes == nil {
+		return ix.Tensor.DimSize(0)
+	}
+	return len(ix.Indexes)
+}
+
+// Len returns the total number of elements in the tensor,
+// taking into account the Indexes via [Rows],
+// as Rows() * cell size.
+func (ix *Indexed) Len() int {
+	rows := ix.NumRows()
+	_, cells := ix.Tensor.RowCellSize()
+	return cells * rows
+}
+
+// DeleteInvalid deletes all invalid indexes from the list.
+// Call this if rows (could) have been deleted from tensor.
+func (ix *Indexed) DeleteInvalid() {
+	if ix.Tensor == nil || ix.Tensor.DimSize(0) <= 0 || ix.Indexes == nil {
+		ix.Indexes = nil
+		return
+	}
+	ni := ix.NumRows()
+	for i := ni - 1; i >= 0; i-- {
+		if ix.Indexes[i] >= ix.Tensor.DimSize(0) {
+			ix.Indexes = append(ix.Indexes[:i], ix.Indexes[i+1:]...)
+		}
+	}
+}
+
+// Sequential sets Indexes to nil, resulting in sequential row-wise access into tensor.
+func (ix *Indexed) Sequential() { //types:add
+	ix.Indexes = nil
+}
+
+// IndexesNeeded is called prior to an operation that needs actual indexes,
+// e.g., Sort, Filter.  If Indexes == nil, they are set to all rows, otherwise
+// current indexes are left as is. Use Sequential, then IndexesNeeded to ensure
+// all rows are represented.
+func (ix *Indexed) IndexesNeeded() { //types:add
+	if ix.Tensor == nil || ix.Tensor.DimSize(0) <= 0 {
+		ix.Indexes = nil
+		return
+	}
+	if ix.Indexes != nil {
+		return
+	}
+	ix.Indexes = make([]int, ix.Tensor.DimSize(0))
+	for i := range ix.Indexes {
+		ix.Indexes[i] = i
+	}
+}
+
+// ExcludeMissing deletes indexes where the values are missing, as indicated by NaN.
+// Uses first cell of higher dimensional data.
+func (ix *Indexed) ExcludeMissing() { //types:add
+	if ix.Tensor == nil || ix.Tensor.DimSize(0) <= 0 {
+		ix.Indexes = nil
+		return
+	}
+	ix.IndexesNeeded()
+	ni := ix.NumRows()
+	for i := ni - 1; i >= 0; i-- {
+		if math.IsNaN(ix.Tensor.FloatRowCell(ix.Indexes[i], 0)) {
+			ix.Indexes = append(ix.Indexes[:i], ix.Indexes[i+1:]...)
+		}
+	}
+}
+
+// Permuted sets indexes to a permuted order.  If indexes already exist
+// then existing list of indexes is permuted, otherwise a new set of
+// permuted indexes are generated
+func (ix *Indexed) Permuted() {
+	if ix.Tensor == nil || ix.Tensor.DimSize(0) <= 0 {
+		ix.Indexes = nil
+		return
+	}
+	if ix.Indexes == nil {
+		ix.Indexes = rand.Perm(ix.Tensor.DimSize(0))
+	} else {
+		rand.Shuffle(len(ix.Indexes), func(i, j int) {
+			ix.Indexes[i], ix.Indexes[j] = ix.Indexes[j], ix.Indexes[i]
+		})
+	}
+}
+
+const (
+	// Ascending specifies an ascending sort direction for tensor Sort routines
+	Ascending = true
+
+	// Descending specifies a descending sort direction for tensor Sort routines
+	Descending = false
+
+	//	Stable specifies using stable, original order-preserving sort, which is slower.
```

Comment:
Your tab formatting is bad here. Also, these bool constants are once again pretty ambiguously named...

Reply:
And it doesn't even look like you use `Stable` and `Unstable`?

Reply:
yes we do use inside stats/group  renamed to StableSort  etc.